File: oolab/src/main/java/agh/ics/oop/model/AbstractWorldMap.java
package agh.ics.oop.model;

import agh.ics.oop.model.util.Boundary;
import agh.ics.oop.model.util.IncorrectPositionException;
import agh.ics.oop.model.util.MapVisualizer;
import java.util.*;

public abstract class AbstractWorldMap implements WorldMap {
    protected final UUID id = UUID.randomUUID();
    protected Vector2d lowerLeft = new Vector2d(Integer.MIN_VALUE, Integer.MIN_VALUE);
    protected Vector2d upperRight = new Vector2d(Integer.MAX_VALUE, Integer.MAX_VALUE);
    protected final Map<Vector2d, Animal> animals = new HashMap<>();
    protected final MapVisualizer visualizer = new MapVisualizer(this);
    protected final List<MapChangeListener> observers = new ArrayList<>();

    public void addObserver(MapChangeListener observer) {
        observers.add(observer);
    }

    public void removeObserver(MapChangeListener observer) {
        observers.remove(observer);
    }

    protected void notifyObservers(String message) {

        synchronized (observers) {
            for (MapChangeListener observer : observers) {
                observer.mapChanged(this, message);
            }
        }
    }

    @Override
    public boolean canMoveTo(Vector2d position) {
        return !(objectAt(position) instanceof Animal);
    }

    @Override
    public boolean place(Animal animal) throws IncorrectPositionException {
        if (canMoveTo(animal.getPosition())) {
            animals.put(animal.getPosition(), animal);
            notifyObservers("Animal placed: " + animal.getPosition());
            return true;
        }
        throw new IncorrectPositionException(animal.getPosition());
    }

    @Override
    public void move(Animal animal, MoveDirection direction) {
        Vector2d oldPosition = animal.getPosition();
        animal.move(direction, this);
        animals.remove(oldPosition);
        animals.put(animal.getPosition(), animal);
        notifyObservers("Animal moved to: " + animal.getPosition() + " from: " + oldPosition);
    }

    @Override
    public boolean isOccupied(Vector2d position) {
        return objectAt(position) != null;
    }

    @Override
    public WorldElement objectAt(Vector2d position) {
        if(animals.get(position) != null) return animals.get(position);
        return null;
    }

    @Override
    public List<WorldElement> getElements() {
        List<WorldElement> elements = new ArrayList<>(animals.values());
        // Add other elements if needed
        return elements;
    }


    @Override
    public Boundary getCurrentBounds() {
        return new Boundary(lowerLeft, upperRight);
    }

    @Override
    public String toString() {
        return visualizer.draw(getCurrentBounds().lowerLeft(), getCurrentBounds().upperRight());
    }

    @Override
    public UUID getID() {
        return id;
    }
}

File: kotlin/src/main/kotlin/module/AbstractWorldMap.kt
package module

import module.util.MapVisualizer
import org.example.module.Animal
import org.example.module.MoveDirection
import org.example.module.Vector2d
import org.example.module.WorldMap
import org.example.module.util.Boundary
import org.example.module.util.IncorrectPositionException
import java.util.*
import kotlin.collections.ArrayList

abstract class AbstractWorldMap : WorldMap {
    private val id: UUID = UUID.randomUUID()
    protected var lowerLeft: Vector2d = Vector2d(Int.MIN_VALUE, Int.MIN_VALUE)
    protected var upperRight: Vector2d = Vector2d(Int.MAX_VALUE, Int.MAX_VALUE)
    protected val animals = mutableMapOf<Vector2d, Animal>()
    protected val visualizer = MapVisualizer(this)


    override fun canMoveTo(position: Vector2d): Boolean =
        objectAt(position)?.let { false } ?: true

    @Throws(IncorrectPositionException::class)
    override fun place(animal: Animal): Boolean {
        if (canMoveTo(animal.position)) {
            animals[animal.position] = animal
            return true
        }
        throw IncorrectPositionException(animal.position)
    }

    override fun move(animal: Animal, direction: MoveDirection) {
        val oldPosition = animal.position
        animal.move(direction, this)
        animals.remove(oldPosition)
        animals[animal.position] = animal
    }

    override fun isOccupied(position: Vector2d): Boolean = animals.containsKey(position)

    override fun objectAt(position: Vector2d): Animal? = animals[position]

    override fun getElements(): List<Animal> = animals.values.toList()

    override fun getCurrentBounds(): Boundary = Boundary(lowerLeft, upperRight)

    override fun getID(): UUID = id

    override fun getOrderedAnimals(): ArrayList<Animal> =
        animals.values.sortedWith(compareBy({ it.position.x }, { it.position.y }))
            .toCollection(ArrayList())

    override fun toString(): String =
        visualizer.draw(getCurrentBounds().lowerLeft, getCurrentBounds().upperRight)

    abstract fun randomFreePostion(vector2d: Vector2d): Vector2d?
}